### ISSUE TITLE
Bump codeeditor package to 0.16.0.

### ISSUE THREAD

**Issue by jasongrout** [jupyterlab/jupyterlab#4376]
See https://gitter.im/jupyterlab/jupyterlab?at=5acf8fba6bbe1d2739dd297a from @afshin.

> @ian-r-rose @jasongrout we made a mistake in the JS releases. @jupyterlab/codeeditor should have been a minor release and not a patch. Steve pointed it out today. So I think we need to do another patch release that fixes the 0.15.x line and also a minor 0.16.0 release for @jupyterlab/codeeditor and anything that imports it

**Comment by ian-r-rose** [comment]
I suppose I have a bit of confusion about semver and pre-1.0 packages. I see two changes to `@jupyterlab/codeeditor` since the last publish: (1) Swapping `@phosphor/virtualdom` for React (implementation detail, no API changes), and (2) Adding `insertLineBreak()`, which extends the API but doesn't break anything.

If this were a 1.0 package, I would expect that to be a minor bump. In this case, I find myself confused, where a minor number bump is semantically major.

**Comment by jasongrout** [comment]
Why does @blink1073 and @afshin think it should be bumped to 0.16.0?

For pre-1.0, technically anything goes. In practice, there are *only* minor and major version bumps: 0.15.3 -> 0.15.4 *is* a "minor" version bump (in that ~ will not do the update, but ^ will).

**Comment by vidartf** [comment]
I'm getting some CI errors on a non-upgraded repo ( a build on master after a non-relevant change) like this:
```
src/common/editor.ts(40,13): error TS2352: Type 'IEditor' cannot be converted to type 'CodeMirrorEditor'.
      Property 'editor' is missing in type 'IEditor'.
    src/common/editor.ts(44,13): error TS2352: Type 'IEditor' cannot be converted to type 'CodeMirrorEditor'.
```

It's basically a casting error from a `@jupyterlab/codeeditor.CodeEditor.IEditor` to a `@jupyterlab/codemirror.CodeMirrorEditor`, since the latest 0.15.x of the latter doesn't implement the interface defined in the latest 0.15.x of the former.

As such, the last 0.15.x of codeeditor should be pulled if possible, and a 0.16.x released instead.

**Comment by vidartf** [comment]
Formulated differently: When adding non-optional parts to an interface/class that is meant to be implemented/subclassed, it is not backwards compatible.

**Comment by jasongrout** [comment]
Thanks @vidartf, that hits the nail on the head. We'll do another 0.15 release of codeeditor with a reversion, then an 0.16 release of codeeditor.

**Comment by jasongrout** [comment]
For reference, the two changes to the codeeditor package in question are:

https://github.com/jupyterlab/jupyterlab/commit/6d121a7ca34562b259bc41cf6c82ddc20d0b8f34#diff-3cf4e3aee7a197ff88eaf118d1ecd4ae

https://github.com/jupyterlab/jupyterlab/commit/6121bee4eda41ec2e089641f0b5f2629a7cd8dc2#diff-3cf4e3aee7a197ff88eaf118d1ecd4ae

My plan is to revert 6121bee4eda41ec2e089641f0b5f2629a7cd8dc2, release codeeditor 0.15.x, commit 6121bee4eda41ec2e089641f0b5f2629a7cd8dc2 again, then release codeeditor 0.16

**Comment by jasongrout** [comment]
Actually, I'll revert and recommit all of the commits in https://github.com/jupyterlab/jupyterlab/pull/4070

**Comment by jasongrout** [comment]
Actually, I think the other commit to codeeditor forces an 0.16 version number as well. It upgrades the react dependency, which means that the codeeditor 0.15.x now has an incompatible react with the other 0.15 packages.

I'll effectively revert that too. 

New plan to do this: I'll go to the 0.31 branch, publish a new 0.15 release of codeeditor there, then publish 0.16 from master.

**Comment by jasongrout** [comment]
I did the following:

1. checked out the 0.31.x branch, then cleaned and rebuilt the js
2. edited the codeeditor package.json to be 0.15.6
3. ran `npm publish` to publish 0.15.6
4. decided it might be better to do things a different way, so `npm unpublish @jupyterlab/codeeditor@0.15.6`
5. After fiddling with things, decided to do it the way I did originally.

then again:
1. checked out the 0.31.x branch, then cleaned and rebuilt the js
2. edited the codeeditor package.json to be 0.15.7
3. ran `npm publish` to publish 0.15.7

**Comment by jasongrout** [comment]
@vidartf - can you check that your error goes away now?

**Comment by jasongrout** [comment]
@ian-r-rose pointed out the instructions for doing a patch release of a package: https://github.com/jupyterlab/jupyterlab/blob/master/RELEASE.md#making-a-patch-release-of-a-javascript-package

I should have just followed those. I'll do *another* release to bump to 0.15.8 the proper way. I'll also unpublish 0.15.7.

**Comment by jasongrout** [comment]
The main problem is that the release is not noted anywhere in the repo.

**Comment by jasongrout** [comment]
I just published codeeditor 0.16.0 from the current jlab master.
